### ISSUE TITLE
Vault Protocol Downgrade Issue

### ISSUE THREAD

**Issue by ccampanale** [hashicorp/consul-template#501]
It appears that CT resolves the address provided for Vault at configuration time which can cause issues later if the leader changes and Vault is configured securely with TLS.

Here's the Vault portion of our CT configuration. We have a dynamic service registration for Vault leadership via Consul so if the leader changes so to will this service registration. 

```hcl
vault {
  address = "https://vault-leader.service.consul:8200"
  token = "*"
  ssl {
    enabled = true
    verify = true
    cert = ""
    ca_cert = "/etc/pki/CA/certs/vault-ca-combo.pem"
  }
}
```

The Vault leader changed at some point over the holidays which caused CT to start complaining about a protocol downgrade (since the Vault TLS instances cannot redirect requests). 

```
2016/01/04 08:06:58 [ERR] (view) vault_token error renewing vault token: redirect would cause protocol downgrade
2016/01/04 08:06:58 [INFO] (view) vault_token errored, retrying in 10s
2016/01/04 08:06:58 [ERR] (runner) watcher reported error: error renewing vault token: redirect would cause protocol downgrade
```

Is there a better way to approach this or would it be possible to investigate why CT would still contact the old instance? It would be nice if CT would always use the provided hostname as opposed to storing the result of the first name resolution. (At least this is my assumption of what is occurring... )

Please let me know if you have any questions I may not have provided here.

Thanks!

**Comment by jefferai** [comment]
Hi @ccampanale ,

Can you check your Vault configuration and ensure that you do not have different TLS settings for the different Vault servers? Vault uses TLS 1.2 by default, and this sounds like one of your servers (that is now the leader) is set to use 1.0 or 1.1 instead.

**Comment by ccampanale** [comment]
Hey @jefferai, thanks for the quick reply!

Here is the Vault configuration shared among the servers in the cluster: 

```json
{
  "backend": {
    "consul": {
      "address": "consulprd1.my.domain:8500",
      "advertise_addr": "http://consulprd1.my.domain",
      "path": "vault",
      "scheme": "http",
      "token": "*"
    }
  },
  "listener": {
    "tcp": {
      "address": "<server's_address>:8200",
      "tls_cert_file": "/path/to/vault.service.consul.crt",
      "tls_key_file": "/path/to/vault.service.consul.key",
      "tls_min_version": "tls10"
    }
  }
}
```

Is it the `tls_min_version` that is causing this? I originally set this because of issues with Curl/Wget as the suggested fix in an issue (can't find it at the moment... but I think it is in the Vault repo). How can I tell what TLS version the current running Vault instances are using?

**Comment by ccampanale** [comment]
I did a few checks with the openssl s_client and each time it appears to have created a session with TLSv1.2 even though Vault is configured to allow TLSv1.0 (This is, I believe, the way this should be working):

Here's what I tested:

`openssl s_client -no_tls1 -no_ssl3 -connect vault-leader.service.consul:8200`

`openssl s_client -no_tls1 -connect vault-leader.service.consul:8200`

`openssl s_client -connect vault-leader.service.consul:8200`

```
New, TLSv1/SSLv3, Cipher is ECDHE-RSA-AES256-GCM-SHA384
Server public key is 2048 bit
Secure Renegotiation IS supported
Compression: NONE
Expansion: NONE
SSL-Session:
    Protocol  : TLSv1.2
    Cipher    : ECDHE-RSA-AES256-GCM-SHA384
    Session-ID: EEB5048A1510E06D11F4C1BB6105E01E49DDFE5A32F1BB9F0C35242974C8B4EA
    Session-ID-ctx:
    Master-Key: 73C9928AE96824CD1794A7EAD280DAA16A51F51D950CD910D7488E3FB116AA607CE6A9F31ACFF28510AD54D2FAAEE074
    Key-Arg   : None
    Krb5 Principal: None
    PSK identity: None
    PSK identity hint: None
    TLS session ticket:
    0000 - 40 52 3b d9 1f 31 d4 23-98 e0 9d e8 a8 94 aa 30   @R;..1.#.......0
    0010 - 08 a3 74 f1 3f 7b ce 5a-41 39 bb 9b 47 97 41 9d   ..t.?{.ZA9..G.A.
    0020 - 51 9f f4 28 8c da b5 8d-ec 60 50 25 65 0c 0f f2   Q..(.....`P%e...
    0030 - c0 37 3a 09 f9 9d e5 9f-59 7e 91 8f 6d 25 08 1d   .7:.....Y~..m%..
    0040 - b0 3b 8e 64 49 13 fa 1e-1d 7c 2c 07 c0 db f2 3f   .;.dI....|,....?
    0050 - 50 79 44 53 bf d0 71 fe-54 2f 9a dd a4 a7 f4 a4   PyDS..q.T/......
    0060 - a4 83 3d 8f 6a a9 a5 1a-db d3 dd eb 62 77 47 b8   ..=.j.......bwG.
    0070 - ac 64 20 c8 79 4f 9e 5d-                          .d .yO.]

    Start Time: 1451926065
    Timeout   : 300 (sec)
    Verify return code: 21 (unable to verify the first certificate)
---
```

I also ran this against each node individually and got the same results.

**Comment by jefferai** [comment]
Hi @ccampanale ,

It's actually not the `tls_min_version` line that is causing the problem, it's this:

>      "advertise_addr": "http://consulprd1.my.domain",

The `advertise_addr` is the value that should be used when redirecting the client to the leader -- for any given Vault server, it is basically an address that says "if you are redirecting a client to _me_, this is the address to give to the client". In an HA setup this may be each node's own address or the address of a shared DNS entry (via service discovery or e.g. an ELB). In your case, based on your config above, my guess is that it should probably be `https://vault-leader.service.consul:8200`, unless your health checks are such that it may take a while for a leadership change to be reflected in Consul, in which case you may want to put the host names directly.

But the key here is that that `advertise_addr` value is `http`, not `https`. So when the client is being redirected from the initial location to the leader, it's being told to go from a secure connection to a non-secure connection, and it's refusing.

**Comment by ccampanale** [comment]
Interesting, so the "advertise_addr" in the Consul backend configuration determines how Vault reacts to external requests? I figured since it was part of the Consul backend configuration that it would simply be for configuration Vault on how to communicate with the Consul cluster. 

I'll give that a whirl and see what happens...

**Comment by jefferai** [comment]
It's slightly different, which is probably the confusing bit. It's actually an aspect of the physical (storage) backend configuration ("backend"), but of which yours is of type "consul". If you look in the documentation, you'll see that it's listed as a "common" option for all backends rather than as a "consul" option.

**Comment by ccampanale** [comment]
Yeah, I actually noticed that is was common when I checked the configuration documentation after your last post. I'm testing the change right now. BTW - I am currently using the node's FQDN for that parameter and I will continue to just with HTTPS. You are right, the vault-leader service registration is determined by a script that runs every fives minutes so there can certainly be a period of time where one node is unregistering as the leader and the newly elected leader registers itself. I will test it this way first and see how it goes. Thanks again!

**Comment by ccampanale** [comment]
Hey @jefferai, thanks again for the help! I do believe that did the trick.

I tested the Vault nodes using the following command pointed directly at each individual node: `vault mounts -address=https://consulprd#.my.domain:8200` Previously (and before this change) it would not be able to forward the request as we saw from the CT logs. So, this command would only work against the leader. I updated the configuration, restarted, and unsealed each node and now I can run this against any node in the cluster. This makes me VERY HAPPY as I should be able to use vault.service.consul without issue!!!

Sorry this ended up not being a CT related issue... I SWEAR I looked through a number of Vault issues and read a bit about it being "unsecure for vault instances to forward requests to the leader when configured with TLS" and that it would be researched for future versions so I wrote this off as something that couldn't be changed yet. This is why I came up with the dynamic registration for `vault-leader.service.consul`. 

So for anyone who may read this in the future the key here is for the `advertise_addr` in the consul backend configuration for Vault to point to the Vault instance's http API port over HTTPS:


```json
{
  "backend": {
    "consul": {
      "address": "consulprd1.my.domain:8500",
      "advertise_addr": "https://consulprd1.my.domain:8200",
      "path": "vault",
      "scheme": "http",
      "token": "*"
    }
  },
  "listener": {
    "tcp": {
      "address": "<server's_address>:8200",
      "tls_cert_file": "/path/to/vault.service.consul.crt",
      "tls_key_file": "/path/to/vault.service.consul.key",
      "tls_min_version": "tls10"
    }
  }
}
```

(Wouldn't that be interesting if somehow this fixes #493 as well? Here's to hoping...) Thanks again for the help!
